File: go/core/interfaces/logger.go
package interfaces

import "context"

// Logger provides structured, leveled logging.
//
// Implementations: zap-backed slog (default), stdlib slog, or any structured
// logger that supports key-value pairs.
//
// All foundation and service code depends on this interface, never on a
// concrete logger. Swap implementations via the logger.New factory.
type Logger interface {
	// Debug logs at DEBUG level with optional key-value pairs.
	Debug(msg string, keysAndValues ...any)

	// Info logs at INFO level with optional key-value pairs.
	Info(msg string, keysAndValues ...any)

	// Warn logs at WARN level with optional key-value pairs.
	Warn(msg string, keysAndValues ...any)

	// Error logs at ERROR level with optional key-value pairs.
	Error(msg string, keysAndValues ...any)

	// With returns a new Logger with the given key-value pairs pre-bound.
	With(keysAndValues ...any) Logger

	// WithContext returns a new Logger that extracts values (e.g. correlation ID)
	// from the context on each log call.
	WithContext(ctx context.Context) Logger
}

File: python/techai_webutils/src/techai_webutils/core/interfaces/logger.py
"""Logger interface for structured logging.

Defines the abstract contract for logger implementations.
Foundation implementations satisfy this interface; consumers depend only on it.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Logger(ABC):
    """Abstract structured logger with level-based logging.

    Implementations: structlog JSON (default), stdlib logging.
    All service code depends on this interface, never on a concrete logger.
    """

    @abstractmethod
    def debug(self, msg: str, **kwargs: object) -> None:
        """Log at DEBUG level with optional key-value context."""
        ...

    @abstractmethod
    def info(self, msg: str, **kwargs: object) -> None:
        """Log at INFO level with optional key-value context."""
        ...

    @abstractmethod
    def warning(self, msg: str, **kwargs: object) -> None:
        """Log at WARNING level with optional key-value context."""
        ...

    @abstractmethod
    def error(self, msg: str, **kwargs: object) -> None:
        """Log at ERROR level with optional key-value context."""
        ...

    @abstractmethod
    def exception(self, msg: str, **kwargs: object) -> None:
        """Log at ERROR level with exception info attached."""
        ...

    @abstractmethod
    def bind(self, **kwargs: object) -> Logger:
        """Return a new Logger with the given key-value pairs pre-bound."""
        ...
